federationapi/queue/queue.go

```diff
@@ -104,28 +104,28 @@ func NewOutgoingQueues(
 	}
 	// Look up which servers we have pending items for and then rehydrate those queues.
 	if !disabled {
-		time.AfterFunc(time.Second*5, func() {
-			serverNames := map[gomatrixserverlib.ServerName]struct{}{}
-			if names, err := db.GetPendingPDUServerNames(context.Background()); err == nil {
-				for _, serverName := range names {
-					serverNames[serverName] = struct{}{}
-				}
-			} else {
-				log.WithError(err).Error("Failed to get PDU server names for destination queue hydration")
+		serverNames := map[gomatrixserverlib.ServerName]struct{}{}
+		if names, err := db.GetPendingPDUServerNames(context.Background()); err == nil {
+			for _, serverName := range names {
+				serverNames[serverName] = struct{}{}
 			}
-			if names, err := db.GetPendingEDUServerNames(context.Background()); err == nil {
-				for _, serverName := range names {
-					serverNames[serverName] = struct{}{}
-				}
-			} else {
-				log.WithError(err).Error("Failed to get EDU server names for destination queue hydration")
+		} else {
+			log.WithError(err).Error("Failed to get PDU server names for destination queue hydration")
+		}
+		if names, err := db.GetPendingEDUServerNames(context.Background()); err == nil {
+			for _, serverName := range names {
+				serverNames[serverName] = struct{}{}
 			}
-			for serverName := range serverNames {
-				if queue := queues.getQueue(serverName); queue != nil {
-					queue.wakeQueueIfNeeded()
-				}
+		} else {
+			log.WithError(err).Error("Failed to get EDU server names for destination queue hydration")
+		}
+		offset := time.Second * 5
+		for serverName := range serverNames {
+			if queue := queues.getQueue(serverName); queue != nil {
+				time.AfterFunc(offset, queue.wakeQueueIfNeeded)
+				offset += time.Second
```

Comment:
This probably isn't good enough if we have a few thousand servers (not entirely unrealistic) as this would wait 33min for 2000 servers. It would be better if it was stochastic over a large but reasonable timeframe (e.g random number between 0-120 and then wake after that many seconds).